File: STMS-main/lib/core/services/signalr_service.dart
import 'package:flutter/foundation.dart';
import 'package:signalr_netcore/signalr_client.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../network/api_constants.dart';

class SignalRService {
  static final SignalRService _instance = SignalRService._internal();
  factory SignalRService() => _instance;
  SignalRService._internal();

  HubConnection? _hubConnection;
  bool _isConnecting = false;

  HubConnection? get connection => _hubConnection;

  Future<void> initConnection() async {
    if (_hubConnection != null && _hubConnection!.state == HubConnectionState.Connected) {
      return;
    }
    if (_isConnecting) return;
    _isConnecting = true;

    try {
      final prefs = await SharedPreferences.getInstance();
      final token = prefs.getString('jwt_token');

      // Swap /api with /hubs/traffic
      final String hubUrl = ApiConstants.baseUrl.replaceAll('/api', '/hubs/traffic');

      final httpOptions = HttpConnectionOptions(
        accessTokenFactory: () async => token ?? '',
      );

      _hubConnection = HubConnectionBuilder()
          .withUrl(hubUrl, options: httpOptions)
          .withAutomaticReconnect()
          .build();

      _hubConnection!.onclose(({error}) {
        debugPrint('SignalR: Connection closed: $error');
      });

      _hubConnection!.onreconnecting(({error}) {
        debugPrint('SignalR: Reconnecting: $error');
      });

      _hubConnection!.onreconnected(({connectionId}) {
        debugPrint('SignalR: Reconnected with connectionId: $connectionId');
      });

      await _hubConnection!.start();
      debugPrint('SignalR: Connection started successfully');
    } catch (e) {
      debugPrint('SignalR: Error starting connection: $e');
    } finally {
      _isConnecting = false;
    }
  }

  Future<void> stopConnection() async {
    try {
      if (_hubConnection != null && _hubConnection!.state == HubConnectionState.Connected) {
        await _hubConnection!.stop();
        debugPrint('SignalR: Connection stopped');
      }
    } catch (e) {
      debugPrint('SignalR: Error stopping connection: $e');
    }
  }

  /// Join a SignalR group for chat
  Future<void> joinTicketRoom(String ticketId) async {
    if (_hubConnection == null || _hubConnection!.state != HubConnectionState.Connected) {
      await initConnection();
    }
    try {
      await _hubConnection?.invoke('JoinTicketRoom', args: [ticketId]);
      debugPrint('SignalR: Joined room $ticketId');
    } catch (e) {
      debugPrint('SignalR: Error joining ticket room: $e');
    }
  }

  /// Listen to message events from the hub
  void registerMessageListener(Function(dynamic) onMessageReceived) {
    _hubConnection?.on('ReceiveMessage', (arguments) {
      if (arguments != null && arguments.isNotEmpty) {
        onMessageReceived(arguments[0]);
      }
    });
  }

  /// Unregister message events from the hub
  void unregisterMessageListener() {
    _hubConnection?.off('ReceiveMessage');
  }
}

File: my-project/src/api/services/signalrService.js
import { HubConnectionBuilder, LogLevel } from "@microsoft/signalr";
import { API_CONFIG } from "../config";

class SignalRService {
  constructor() {
    this.connection = null;
    this.listeners = new Map();
  }

  async startConnection() {
    if (this.connection) return this.connection;

    const token = localStorage.getItem("token");
    // Swap /api with /hubs/traffic
    const hubUrl = API_CONFIG.BASE_URL.replace("/api", "/hubs/traffic");

    this.connection = new HubConnectionBuilder()
      .withUrl(hubUrl, {
        accessTokenFactory: () => token || "",
      })
      .withAutomaticReconnect([0, 2000, 5000, 10000, 30000]) // Custom backoff retry intervals
      .configureLogging(LogLevel.Warning)
      .build();

    this.connection.onclose((error) => {
      console.warn("SignalR: Connection closed.", error);
    });

    this.connection.onreconnecting((error) => {
      console.warn("SignalR: Connection lost. Reconnecting...", error);
    });

    this.connection.onreconstructed((connectionId) => {
      console.info("SignalR: Connection re-established. Connection ID:", connectionId);
    });

    try {
      await this.connection.start();
      console.info("SignalR: Connected successfully!");
    } catch (err) {
      console.error("SignalR: Error establishing connection.", err);
      // Retry in 5 seconds
      setTimeout(() => this.startConnection(), 5000);
    }

    return this.connection;
  }

  async stopConnection() {
    if (!this.connection) return;
    try {
      await this.connection.stop();
      this.connection = null;
      console.info("SignalR: Connection stopped.");
    } catch (err) {
      console.error("SignalR: Error stopping connection.", err);
    }
  }

  async joinTicketRoom(ticketId) {
    const conn = await this.startConnection();
    if (conn.state === "Connected") {
      try {
        await conn.invoke("JoinTicketRoom", ticketId);
        console.info(`SignalR: Joined chat room ${ticketId}`);
      } catch (err) {
        console.error("SignalR: Error joining ticket room.", err);
      }
    }
  }

  registerMessageListener(onMessageReceived) {
    if (!this.connection) return;
    this.connection.on("ReceiveMessage", onMessageReceived);
  }

  unregisterMessageListener(onMessageReceived) {
    if (!this.connection) return;
    this.connection.off("ReceiveMessage", onMessageReceived);
  }
}

const signalrService = new SignalRService();
export default signalrService;
